### PR TITLE
Add high-concurrency Livy support for parallel statement execution

### PR DESCRIPTION
**Pull request by sdebruyn**
## Summary

- Add [Fabric high-concurrency Livy API](https://learn.microsoft.com/en-us/fabric/data-engineering/high-concurrency-livy?WT.mc_id=MVP_310840) support so each dbt thread gets its own REPL inside a shared underlying Livy session — statements execute in parallel instead of queuing FIFO
- HC is the only session mode — the legacy single-session path has been removed entirely
- Deterministic `sessionTag` from `(workspace_id, lakehouse_id)` lets Fabric snap-attach REPLs onto a warm session across dbt invocations
- HC session cleanup via dbt's connection manager `close()` path — no `atexit` handlers
- Both adapters now use HC Livy: FabricSpark for all SQL execution, Fabric DW for Python model execution

Based on upstream contribution: [microsoft/dbt-fabricspark#186](https://github.com/microsoft/dbt-fabricspark/pull/186). Clean-room implementation adapted to our `FabricApiClient` + `HighConcurrencyLivySession` + `FabricSparkCursor` architecture.

Closes #231

## Changed files

| File | Change |
|---|---|
| `fabric_api_client.py` | HC API methods only (acquire, poll, submit/get/cancel statement, delete); legacy single-session methods removed |
| `fabric_hc_livy_session.py` | **New** — `HighConcurrencyLivySession` with REPL lifecycle, polling, and best-effort cleanup on failure |
| `fabric_livy_session.py` | Reduced to shared dataclasses (`LivySessionResult`, `LivySubmissionResult`); `LivySession` class removed |
| `fabric_livy_helper.py` | Switched to `HighConcurrencyLivySession` with thread-local storage for per-thread REPL isolation |
| `fabricspark_connection_manager.py` | Always creates HC session (no legacy fallback) |
| `fabricspark_cursor.py` | Routes cancel through session interface (`cancel_statement`) |
| `fabricspark_connection.py` | Typed for `HighConcurrencyLivySession` only |
| `docs/lakehouse.md` | HC architecture docs, Mermaid diagram, performance considerations |
| `docs/comparison-dbt-fabricspark.md` | Updated comparison table |

## Test plan

- [ ] Run FabricSpark integration tests (`/test-de`)
- [ ] Run Fabric DW integration tests with Python models (`/test-dw`)
- [ ] Verify `threads > 5` creates multiple underlying sessions

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: src/dbt/adapters/fabricspark/fabricspark_cursor.py [hunk 1]
```diff
@@ -142,7 +142,7 @@ def cancel(self) -> None:
         if self._connection is None:
             return
         if self._statement_id is not None and self._result is None:
-            self.get_livy_session()._fabric_api_client.cancel_livy_statement(self._statement_id)
+            self.get_livy_session().cancel_statement(self._statement_id)
             self._statement_id = None
```
[the remark this answers is not in the document]
Comment (sdebruyn) [reply src/dbt/adapters/fabricspark/fabricspark_cursor.py:147]: Fixed in c2fac39 — the unit test now asserts `livy_session.cancel_statement(42)` instead of the old direct API client path.

File: src/dbt/adapters/fabricspark/fabricspark_cursor.py [hunk 2]
```diff
@@ -142,7 +142,7 @@ def cancel(self) -> None:
         if self._connection is None:
             return
         if self._statement_id is not None and self._result is None:
-            self.get_livy_session()._fabric_api_client.cancel_livy_statement(self._statement_id)
+            self.get_livy_session().cancel_statement(self._statement_id)
```
[the remark this answers is not in the document]
Comment (sdebruyn) [reply src/dbt/adapters/fabricspark/fabricspark_cursor.py:146]: Already addressed in d0e829e — since we removed the non-HC path, the cursor annotation is now `HighConcurrencyLivySession` (no union needed).

File: src/dbt/adapters/fabricspark/fabricspark_credentials.py [hunk 3]
```diff
@@ -5,6 +5,8 @@
 
 @dataclass
 class FabricSparkCredentials(BaseFabricCredentials):
+    high_concurrency: bool = True
+
     @property
     def type(self):
         return "fabricspark"
```
[the remark this answers is not in the document]
Comment (sdebruyn) [reply src/dbt/adapters/fabricspark/fabricspark_credentials.py:10]: No longer applicable — `high_concurrency` was removed in d0e829e. HC is now the only mode, so there's nothing to distinguish in connection identity.

File: src/dbt/adapters/fabric/fabric_hc_livy_session.py [hunk 4]
```diff
@@ -0,0 +1,275 @@
...
+        last_exception: Exception | None = None
+
+        for attempt in range(1, max_attempts + 1):
+            try:
+                body = self._fabric_api_client.acquire_hc_session(tag)
+                break
+            except FabricApiError as e:
+                is_transient = e.status_code == 404 or 500 <= e.status_code < 600
+                if not is_transient or attempt == max_attempts:
+                    raise
```
[the remark this answers is not in the document]
Comment (sdebruyn) [reply src/dbt/adapters/fabric/fabric_hc_livy_session.py:103]: Fixed in c2fac39 — `wait_for_session_ready()` now catches `requests.exceptions.ConnectionError`, `Timeout`, `ChunkedEncodingError`, and `json.JSONDecodeError` alongside `FabricApiError`, matching the singleton `LivySession` resilience.

File: src/dbt/adapters/fabric/fabric_hc_livy_session.py [hunk 5]
```diff
@@ -0,0 +1,275 @@
...
+                )
+
+            try:
+                body = self._fabric_api_client.get_hc_session(self._state.hc_id)
+                consecutive_errors = 0
+            except FabricApiError as e:
+                consecutive_errors += 1
+                if consecutive_errors >= self._MAX_CONSECUTIVE_TRANSIENT_ERRORS:
+                    raise
+                logger.warning(
```
[the remark this answers is not in the document]
Comment (sdebruyn) [reply src/dbt/adapters/fabric/fabric_hc_livy_session.py:152]: Fixed in c2fac39 — `_poll_until_idle()` now catches the same `_TRANSIENT_EXCEPTIONS` tuple (shared with `wait_for_session_ready`).

File: src/dbt/adapters/fabric/fabric_hc_livy_session.py [hunk 6]
```diff
@@ -0,0 +1,275 @@
...
+from dataclasses import dataclass
+from typing import Any
+
+from dbt.adapters.events.logging import AdapterLogger
+from dbt.adapters.fabric.fabric_api_client import FabricApiClient, FabricApiError
+from dbt.adapters.fabric.fabric_livy_session import LivySessionResult
+
+logger = AdapterLogger("fabricspark")
+
+_ACQUIRING_STATES = frozenset({"NotStarted", "starting", "AcquiringHighConcurrencySession"})
```
[the remark this answers is not in the document]
Comment (sdebruyn) [reply src/dbt/adapters/fabric/fabric_hc_livy_session.py]: Removed in c2fac39. The polling logic only checks for terminal bad states and the `Idle` target state — acquiring states are implicitly handled by continuing to poll.

File: src/dbt/adapters/fabric/fabric_hc_livy_session.py [hunk 7]
```diff
@@ -0,0 +1,275 @@
...
+class HighConcurrencyLivySession:
+    """One HC REPL per dbt thread — same interface as ``LivySession``.
+
+    Acquires an HC session via ``POST /highConcurrencySessions``, polls until
+    the underlying Livy session is idle and a REPL is allocated, then submits
+    statements through the REPL endpoint.
+
+    Cleanup happens via ``close()`` which DELETEs the HC session, freeing the
+    REPL slot immediately.
+    """
```
[the remark this answers is not in the document]
Comment (sdebruyn) [reply src/dbt/adapters/fabric/fabric_hc_livy_session.py:57]: The HC session path is covered by integration tests (TestValidateConnectionSpark, TestConcurrencyFabricSpark, TestSimpleMaterializationsSpark all pass with HC). Adding mock-based unit tests for HC internals would duplicate what the integration tests already verify against real Fabric infrastructure.

File: src/dbt/adapters/fabric/fabric_api_client.py [hunk 8]
```diff
@@ -538,3 +538,82 @@ def cancel_livy_statement(self, statement_id: int) -> str:
...
+        Returns:
+            The JSON response body containing at least ``id`` and ``state``.
+        """
+        url = self.get_livy_base_api_uri() + "/highConcurrencySessions"
+        body: dict[str, Any] = {
+            "sessionTag": session_tag,
+            "name": self._credentials.livy_session_name,
+        }
+        response = self._api_post(url, body)
+        return response.json()
```
[the remark this answers is not in the document]
Comment (sdebruyn) [reply src/dbt/adapters/fabric/fabric_api_client.py]: Same reasoning as above — the HC API methods are exercised end-to-end by the integration tests. The existing unit tests for `FabricApiClient` cover the non-HC methods; adding mock-based coverage for the HC endpoints would test HTTP plumbing rather than behavior.

File: src/dbt/adapters/fabricspark/fabricspark_connection_manager.py [hunk 9]
```diff
@@ -48,16 +48,19 @@ def open(cls, connection: Connection) -> Connection:
             logger.debug("Connection is already open, skipping open.")
             return connection
 
+        credentials = connection.credentials
+
         def connect() -> FabricSparkConnection:
-            livy_session = LivySession(cls.get_fabric_api_client(connection.credentials))
+            api_client = cls.get_fabric_api_client(credentials)
+            livy_session = HighConcurrencyLivySession(api_client)
             livy_session.wait_for_session_ready()
```
[the remark this answers is not in the document]
Comment (sdebruyn) [reply src/dbt/adapters/fabricspark/fabricspark_connection_manager.py:56]: Fixed in fd02d95 — PR description updated to reflect HC-only mode (no `high_concurrency` flag, no legacy fallback).

File: src/dbt/adapters/fabric/fabric_livy_helper.py [hunk 10]
```diff
@@ -19,7 +20,7 @@ def __init__(self, parsed_model: dict, credential: FabricCredentials) -> None:
         )
 
         if not self._livy_session:
-            self._livy_session = LivySession(fabric_api_client)
+            self._livy_session = HighConcurrencyLivySession(fabric_api_client)
```
[the remark this answers is not in the document]
Comment (sdebruyn) [reply src/dbt/adapters/fabric/fabric_livy_helper.py:26]: Fixed in fd02d95 — `_livy_session` is now stored in `threading.local()` so each thread gets its own `HighConcurrencyLivySession` instance with independent REPL state.

File: src/dbt/adapters/fabric/fabric_hc_livy_session.py [hunk 11]
```diff
@@ -0,0 +1,277 @@
...
+        else:
+            assert last_exception is not None
+            raise last_exception
+
+        hc_id = body.get("id")
+        if not hc_id:
+            raise RuntimeError(f"HC acquire response missing 'id': {body}")
+
+        self._state.hc_id = str(hc_id)
+        self._poll_until_idle()
```
[the remark this answers is not in the document]
Comment (sdebruyn) [reply src/dbt/adapters/fabric/fabric_hc_livy_session.py]: Fixed in fd02d95 — `wait_for_session_ready()` now wraps `_poll_until_idle()` in try/except and does a best-effort `delete_hc_session(hc_id)` before re-raising, preventing REPL slot leaks.

File: docs/lakehouse.md [hunk 12]
```diff
@@ -45,37 +45,36 @@ The FabricSpark adapter does not use the [`host`](configuration.md#host) option
...
     end
-    Adapter->>Livy API: POST /sessions/{id}/statements
-    Livy API->>Spark Session: Execute Spark SQL
+    Adapter->>HC Livy API: POST /sessions/{id}/repls/{replId}/statements
+    HC Livy API->>Spark Session: Execute Spark SQL (in REPL)
     loop Poll every 3 seconds
-        Adapter->>Livy API: GET /statements/{id}
-        Livy API-->>Adapter: Status + results (when done)
+        Adapter->>HC Livy API: GET /statements/{id}
+        HC Livy API-->>Adapter: Status + results (when done)
```
[the remark this answers is not in the document]
Comment (sdebruyn) [reply docs/lakehouse.md]: Fixed in fd02d95 — diagram now shows `POST /highConcurrencySessions/{id}/repls/{replId}/statements` and the matching GET path for polling.

File: src/dbt/adapters/fabric/fabric_hc_livy_session.py [hunk 13]
```diff
@@ -0,0 +1,277 @@
...
+                return
+
+            time.sleep(self._POLLING_INTERVAL)
+
+    def _ensure_repl(self) -> None:
+        """Re-acquire this thread's HC session if it was marked dead."""
+        if self._state.is_dead or self._state.hc_id is None:
+            logger.debug("HC REPL marked stale — re-acquiring")
+            self.wait_for_session_ready()
+
```
[the remark this answers is not in the document]
Comment (sdebruyn) [reply src/dbt/adapters/fabric/fabric_hc_livy_session.py:181]: Fixed in fd02d95 — `_ensure_repl()` now does a best-effort `delete_hc_session()` on the old `hc_id` and resets state before re-acquiring.
